File: packages/sdk/src/resources/user-info.ts
import type { LinkOptions } from '@/config';
import { BaseResource } from '@/resources/base';
import type { IUserInfoResource } from '@/resources/interfaces';
import type { UserInfo } from '@/types/index';
import { z } from 'zod';

const rollingSpendLimitSchema = z.object({
  limit: z.number().int().nullable(),
  used: z.number().int(),
  remaining: z.number().int().nullable(),
});

const agentWalletSpendLimitsSchema = z.object({
  per_transaction: z.object({
    limit: z.number().int().nullable(),
  }),
  daily: rollingSpendLimitSchema,
  thirty_day: rollingSpendLimitSchema,
});

const agentWalletStepUpSchema = z.object({
  status: z.enum([
    'not_required',
    'ssn_verification',
    'identity_verification',
    'contact_support',
    'complete',
  ]),
});

const userInfoSchema = z
  .looseObject({
    email: z.string().nullable().optional(),
    name: z.string().nullable().optional(),
    first_name: z.string().nullable().optional(),
    last_name: z.string().nullable().optional(),
    phone: z.string().nullable().optional(),
    agent_wallet_spend_limits: agentWalletSpendLimitsSchema.optional(),
    agent_wallet_step_up: agentWalletStepUpSchema.optional(),
  })
  .transform(
    ({
      email,
      name,
      first_name,
      last_name,
      phone,
      agent_wallet_spend_limits,
      agent_wallet_step_up,
    }) => ({
      email: email ?? null,
      name: name ?? null,
      first_name: first_name ?? null,
      last_name: last_name ?? null,
      phone: phone ?? null,
      ...(agent_wallet_spend_limits === undefined
        ? {}
        : { agent_wallet_spend_limits }),
      ...(agent_wallet_step_up === undefined ? {} : { agent_wallet_step_up }),
    }),
  );

export class UserInfoResource
  extends BaseResource
  implements IUserInfoResource
{
  constructor(options: LinkOptions) {
    super(options, '/userinfo');
  }

  async retrieve(): Promise<UserInfo> {
    const { status, data, rawBody } = await this.apiFetch({
      method: 'GET',
      url: this.endpoint,
    });

    if (status < 200 || status >= 300) {
      this.throwApiError('retrieve user info', status, data, rawBody);
    }

    return this.parseResponse('retrieve user info', status, () =>
      userInfoSchema.parse(data),
    );
  }
}

File: packages/sdk-go/user_info.go
package link

import (
	"context"
	"net/http"
)

// UserInfoResource provides Link user-information operations.
type UserInfoResource struct {
	base *baseResource
}

// Retrieve returns information about the authenticated Link user.
func (r *UserInfoResource) Retrieve(ctx context.Context) (*UserInfo, error) {
	response, err := r.base.fetch(ctx, http.MethodGet, r.base.baseURL+"/userinfo", nil, nil)
	if err != nil {
		return nil, err
	}
	if response.status < 200 || response.status >= 300 {
		return nil, newAPIError("retrieve user info", response.status, response.data, response.rawBody)
	}
	var wire struct {
		Email                  *string                             `json:"email"`
		Name                   *string                             `json:"name"`
		FirstName              *string                             `json:"first_name"`
		LastName               *string                             `json:"last_name"`
		Phone                  *string                             `json:"phone"`
		AgentWalletSpendLimits *AgentWalletSpendLimits             `json:"agent_wallet_spend_limits,omitempty"`
		AgentWalletStepUp      *AgentWalletVerificationRequirement `json:"agent_wallet_step_up,omitempty"`
	}
	if err := decodeResponse("retrieve user info", response, &wire); err != nil {
		return nil, err
	}
	return &UserInfo{
		Email:                              wire.Email,
		Name:                               wire.Name,
		FirstName:                          wire.FirstName,
		LastName:                           wire.LastName,
		Phone:                              wire.Phone,
		AgentWalletSpendLimits:             wire.AgentWalletSpendLimits,
		AgentWalletVerificationRequirement: wire.AgentWalletStepUp,
	}, nil
}
